src/mbgl/layout/symbol_instance.cpp

```diff
@@ -18,36 +18,79 @@ const Shaping& getAnyShaping(const ShapedTextOrientations& shapedTextOrientation
+                                                   optional<PositionedIcon> shapedIcon,
+                                                   const style::SymbolLayoutProperties::Evaluated& layout,
+                                                   const float layoutTextSize,
+                                                   const style::SymbolPlacementType textPlacement,
+                                                   const std::array<float, 2>& textOffset,
+                                                   const GlyphPositions& positions) : line(std::move(line_)) {
+    // Create the quads used for rendering the icon and glyphs.
+    if (shapedIcon) {
+        iconQuad = getIconQuad(*shapedIcon, layout, layoutTextSize, shapedTextOrientations.horizontal);
+    }
+
+    bool singleLineInitialized = false;
+    const auto initHorizontalGlyphQuads = [&] (SymbolQuads& quads, const Shaping& shaping) {
+        if (!shapedTextOrientations.singleLine) {
+            quads = getGlyphQuads(shaping, textOffset, layout, textPlacement, positions);
+            return;
+        }
+        if (!singleLineInitialized) {
+            rightJustifiedGlyphQuads = getGlyphQuads(shaping, textOffset, layout, textPlacement, positions);
+            singleLineInitialized = true;
+        }
+    };
+
+    if (shapedTextOrientations.right) {
+        initHorizontalGlyphQuads(rightJustifiedGlyphQuads, shapedTextOrientations.right);
+    }
+
+    if (shapedTextOrientations.center) {
+        initHorizontalGlyphQuads(centerJustifiedGlyphQuads, shapedTextOrientations.center);
+    }
+
+    if (shapedTextOrientations.left) {
+        initHorizontalGlyphQuads(leftJustifiedGlyphQuads, shapedTextOrientations.left);
+    }
+
+    if (shapedTextOrientations.vertical) {
+        verticalGlyphQuads = getGlyphQuads(shapedTextOrientations.vertical, textOffset, layout, textPlacement, positions);
+    }
+}
+
+bool SymbolInstanceSharedData::empty() const {
+    return rightJustifiedGlyphQuads.empty() && centerJustifiedGlyphQuads.empty() && leftJustifiedGlyphQuads.empty() && verticalGlyphQuads.empty();
+}
+
 SymbolInstance::SymbolInstance(Anchor& anchor_,
-                               GeometryCoordinates line_,
+                               std::shared_ptr<SymbolInstanceSharedData> sharedData_,
                                const ShapedTextOrientations& shapedTextOrientations,
                                optional<PositionedIcon> shapedIcon,
-                               const SymbolLayoutProperties::Evaluated& layout,
-                               const float layoutTextSize,
                                const float textBoxScale_,
                                const float textPadding,
                                const SymbolPlacementType textPlacement,
-                               const std::array<float, 2> textOffset_,
+                               const std::array<float, 2>& textOffset_,
```

Comment:
Why it is fine here to store reference to textOffset (std::array) and it is not fine to store reference to GeometryCoordinates line_ (std::vector)?

Reply:
textOffset_ is needed after vertices are created, would complicate code a bit. I tried to move biggest 'intermediate' arrays that are shared between instances and can be released once vertices are created.